### ISSUE TITLE
Hint Mode behaves differently for section variable

### ISSUE THREAD

**Issue by coqbot** [coq/coq#5735]
Note: the issue was created automatically with bugzilla2github tool

Original bug ID: BZ#5735
From: @robbertkrebbers
Reported version: 8.7+beta1
CC:   @mattam82

**Comment by RalfJung** [comment]
Behavior changed in master, now both definitions *fail*. In fact, why should they succeed? There are no instances for the lists...

**Comment by RalfJung** [comment]
The following modified example however still behaves exactly as you described:
```coq
Require Import Utf8 Relation_Definitions.

Class Equiv A := equiv: relation A.
Hint Mode Equiv ! : typeclass_instances.

Class Lookup (K A M : Type) := lookup: K → M → option A.
Hint Mode Lookup ! - - : typeclass_instances.
Hint Mode Lookup - - ! : typeclass_instances.

Parameter list_equiv : ∀ A, Equiv A → Equiv (list A).
Parameter option_equiv : ∀ A, Equiv A → Equiv (option A).
Parameter list_lookup : ∀ A, Lookup nat A (list A).

Existing Instance list_equiv.
Existing Instance option_equiv.
Existing Instance list_lookup.

Set Typeclasses Debug.
Lemma list_equiv_lookup {A} `{Equiv A} (l k : list A) :
  equiv l k ↔ ∀ i, equiv (lookup i l) (lookup i k).
Admitted.

Section test.
  Context {A} `{Equiv A}.

  Lemma list_equiv_lookup_2 (l k : list A) :
    equiv l k ↔ ∀ i, equiv (lookup i l) (lookup i k).
  (*
Error:
Unable to satisfy the following constraints:
In environment:
A : Type
H : Equiv A
l, k : list A
i : ?K

?Equiv : "Equiv (option ?A)"

?A : "Type"

?Lookup : "Lookup ?K ?A (list A)"

?Lookup0 : "Lookup ?K ?A (list A)"

?K : "Type"
  *)
End test.
```

The failing log is
```
<infomsg>Debug: 1: looking for (Equiv (list A)) without backtracking</infomsg>
<infomsg>Debug: 1.1: simple apply list_equiv on (Equiv (list A)), 1 subgoal(s)</infomsg>
<infomsg>Debug: 1.1-1 : (Equiv A)</infomsg>
<infomsg>Debug: 1.1-1: looking for (Equiv A) without backtracking</infomsg>
<infomsg>Debug: 1.1-1.1: exact H on (Equiv A), 0 subgoal(s)</infomsg>
<infomsg>Debug: 1: looking for (Equiv (option ?A)) with backtracking</infomsg>
<infomsg>Debug: 1.1: simple apply option_equiv on (Equiv (option ?A)), 1 subgoal(s)</infomsg>
<infomsg>Debug: 1.1-1 : (Equiv ?A)</infomsg>
<infomsg>Debug: 1.1-1: looking for (Equiv ?A) with backtracking</infomsg>
<infomsg>Debug: 1.1-1: no match for (Equiv ?A), 0 possibilities</infomsg>
<infomsg>Debug: 1: looking for (Equiv (option ?A)) with backtracking</infomsg>
<infomsg>Debug: 1.1: simple apply option_equiv on (Equiv (option ?A)), 1 subgoal(s)</infomsg>
<infomsg>Debug: 1.1-1 : (Equiv ?A)</infomsg>
<infomsg>Debug: 1.1-1: looking for (Equiv ?A) with backtracking</infomsg>
<infomsg>Debug: 1.1-1: no match for (Equiv ?A), 0 possibilities</infomsg>
<infomsg>Debug: 1: looking for (Equiv (option ?A)) with backtracking</infomsg>
<infomsg>Debug: 1.1: simple apply option_equiv on (Equiv (option ?A)), 1 subgoal(s)</infomsg>
<infomsg>Debug: 1.1-1 : (Equiv ?A)</infomsg>
<infomsg>Debug: 1.1-1: looking for (Equiv ?A) with backtracking</infomsg>
<infomsg>Debug: 1.1-1: no match for (Equiv ?A), 0 possibilities</infomsg>
```
Seems like with section variables, it *first* checks the `Mode` and then considers the section variables as instances?

**Comment by maximedenes** [comment]
Since MPI as about to officially get premium support ;) , I'll monitor this to make sure we fix it, and will provide some help if needed.

**Comment by RalfJung** [comment]
> In general, we would like to enable Hint Mode for all type classes so make sure instance search does not diverge in unexpected ways. Because of this bug we cannot enable it for Equiv, the type class that is used most throughout the Iris and std++ development.

I'm not even sure what the intended behavior is here. Are you saying you *want* section variables to ignore `Hint Mode`? That also seems problematic. Isn't the actual bug that parameters (stuff in the current context) is applied despite `Hint Mode`?

**Comment by robbertkrebbers** [comment]
The behavior of the version outside the section is correct; the one inside the section should behave the same, and currently is doing something arbitrary.

`Hint Mode` should not be ignored, it's also not ignored by the one outside. For example, if you do:

```
Lemma list_equiv_lookup {A} `{Equiv A} l k :
  equiv l k ↔ ∀ i, equiv (lookup i l) (lookup i k).
```

It will rightfully fail.

**Comment by RalfJung** [comment]
I had to put these next to each other to even see the difference:
```
Lemma list_equiv_lookup {A} `{Equiv A} (l k : list A) :
  equiv l k ↔ ∀ i, equiv (lookup i l) (lookup i k).
```
Okay, so I retract my attempts at explaining what does on. I don't understand at all why anything happens the way it does here.^^

In the succeeding case, the log is
```
Debug: 1: looking for (Equiv (list A)) without backtracking
Debug: 1.1: simple apply list_equiv on (Equiv (list A)), 1 subgoal(s)
Debug: 1.1-1 : (Equiv A)
Debug: 1.1-1: looking for (Equiv A) without backtracking
Debug: 1.1-1.1: exact H on (Equiv A), 0 subgoal(s)
Debug: 1: looking for (Equiv (option ?A)) with backtracking
Debug: 1.1: simple apply option_equiv on (Equiv (option ?A)), 1 subgoal(s)
Debug: 1.1-1 : (Equiv ?A)
Debug: 1.1-1: looking for (Equiv ?A) with backtracking
Debug: 1.1-1.1: exact H on (Equiv ?A), 0 subgoal(s)
Debug: 2: looking for (Lookup ?K A (list A)) with backtracking
Debug: 2.1: simple apply list_lookup on (Lookup ?K A (list A)), 0 subgoal(s)
Debug: 3: looking for (Lookup nat A (list A)) without backtracking
Debug: 3.1: simple apply list_lookup on (Lookup nat A (list A)), 0 subgoal(s)
```
So it actually apply `H` to `Equiv ?A` -- looks a lot like ignoring `Hint Mode` to me!

**Comment by maximedenes** [comment]
@mattam82 do you agree this is a bug? If yes, what should we do to fix it?

**Comment by mattam82** [comment]
There is a bug indeed: the outside one should be failing:
By typechecking the forall i. equiv (lookup ...) we get constraints `Lookup ?A (list A) ?B` and `Equiv (option ?B)`, the later being tried first. It generates Equiv ?B which doesn't match the modes and fails.
The one outside allows instantiation of this Equiv ?B nonetheless. That's because it comes from the local hint database which doesn't inherit the hint modes. class_tactics:L567 should create an emptydatabase inheriting the hints from the typeclass_instances hint database in the only_classes mode at l=east.

**Comment by mattam82** [comment]
To make this work overall, we'd need to postpone constraint Equiv ?A instead of failing right away.

**Comment by RalfJung** [comment]
This seems closely related to https://github.com/coq/coq/issues/5752: that issue probably explains why the first lemma works. The same bug does not arise for section variables, which might be why it fails then.

**Comment by RalfJung** [comment]
> By typechecking the forall i. equiv (lookup ...) we get constraints Lookup ?A (list A) ?B and Equiv (option ?B), the later being tried first. It generates Equiv ?B which doesn't match the modes and fails.
The one outside allows instantiation of this Equiv ?B nonetheless.

Reading this, it seems like there are two bugs here:
* The difference in behavior between the two lemmas is caused by https://github.com/coq/coq/issues/5752. Fixing that issue will "fix" the inconsistency @robbertkrebbers complained about -- both lemmas will fail.
* However, really we'd like both to succeed: just because `Hint Mode` blocks progress in one goal now, doesn't mean search should be aborted; progress made elsewhere can unblock the goal in question.

**Comment by maximedenes** [comment]
I confirm that https://github.com/coq/coq/pull/9996 fixes the discrepancy, by making both examples fail. Making it succeed will require to rework significantly the search algorithm. I can try to do it, but will need to look more closely at it first.

**Comment by maximedenes** [comment]
What remains is exactly #9058, so I'm closing this one. @RalfJung feel free to use the Consortium label there too.

**Comment by RalfJung** [comment]
Indeed that PR landing has uncovered a few places in our various developments where we did some crazy unconstrained TC search.

> What remains is exactly #9058, so I'm closing this one. @RalfJung feel free to use the Consortium label there too.

All right, done.

**Comment by robbertkrebbers** [comment]
We tried to enable `Hint Mode` for `Equiv` in std++ and Iris with Coq 8.13.2, but there still appears to be a problem.

From what I can tell, in https://github.com/coq/coq/issues/5735#issuecomment-496476728 @maximedenes said that #9058 should make sure that **both examples succeed**. However, even with #9058, both examples still fail.

For the record, here's self-contained code:

```coq
Require Import Utf8 Relation_Definitions.

Class Equiv A := equiv: relation A.
Hint Mode Equiv ! : typeclass_instances.

Class Lookup (K A M : Type) := lookup: K → M → option A.
Hint Mode Lookup ! - - : typeclass_instances.
Hint Mode Lookup - - ! : typeclass_instances.

Parameter list_equiv : ∀ A, Equiv A → Equiv (list A).
Parameter option_equiv : ∀ A, Equiv A → Equiv (option A).
Parameter list_lookup : ∀ A, Lookup nat A (list A).

Existing Instance list_equiv.
Existing Instance option_equiv.
Existing Instance list_lookup.

Set Typeclasses Debug.
(* fails *)
Fail Lemma list_equiv_lookup {A} `{Equiv A} (l k : list A) :
  equiv l k ↔ ∀ i, equiv (lookup i l) (lookup i k).

Section test.
  Context {A} `{Equiv A}.

  (* fails *)
  Fail Lemma list_equiv_lookup_2 (l k : list A) :
    equiv l k ↔ ∀ i, equiv (lookup i l) (lookup i k).
  (*
Error:
Unable to satisfy the following constraints:
In environment:
A : Type
H : Equiv A
l, k : list A
i : ?K

?Equiv : "Equiv (option ?A)"

?A : "Type"

?Lookup : "Lookup ?K ?A (list A)"

?Lookup0 : "Lookup ?K ?A (list A)"

?K : "Type"
  *)
End test.
```

What's the best way forward? Should I open a new issue?

**Comment by RalfJung** [comment]
>  Should I open a new issue?

I would say yes, discussions on closed issues tend to get lost.

**Comment by robbertkrebbers** [comment]
> > Should I open a new issue?
> 
> I would say yes, discussions on closed issues tend to get lost.

Sure, let's continue in https://github.com/coq/coq/issues/14441
